src/libexpr/include/nix/expr/value.hh

```diff
@@ -643,6 +748,65 @@ protected:
     {
         setUntaggablePayload<pdPath>(path.accessor, path.path);
     }
+
+    void setStorage(Failed * failed) noexcept
+    {
+        setSingleDWordPayload<tFailed>(std::bit_cast<PackedPointer>(failed));
+    }
+
+    ValueStorage() {}
+
+    ValueStorage(const ValueStorage & v)
+    {
+        *this = v;
+    }
+
+    /**
+     * Copy a value. This is not allowed to be a thunk to avoid
+     * accidental work duplication.
+     */
+    ValueStorage & operator=(const ValueStorage & v)
+    {
+        auto p1_ = v.p1;
+        auto p0_ = v.p0.load(std::memory_order_acquire);
+        auto pd = static_cast<PrimaryDiscriminator>(p0_ & discriminatorMask);
+        if (pd == pdThunk || pd == pdPending || pd == pdAwaited) {
+            printError("UNFINISHED %x %08x %08x", this, p0_, p1_);
+            unreachable();
+        }
+        finish(p0_, p1_);
+        return *this;
+    }
+
+public:
+
+    inline void reset()
+    {
+        p1 = 0;
+        p0.store(0, std::memory_order_relaxed);
+    }
+
+    /// Only used for testing.
+    inline void mkBlackhole()
+    {
+        p0.store(pdPending, std::memory_order_relaxed);
+    }
```

Comment:
If it's only used for testing, is there a way to define this in a way such that it is actually only used in testing? To prevent accidentally using it somewhere it shouldn't be used (e.g. outside of tests)